src/emulator/types.ts

```diff
@@ -57,20 +76,70 @@ export interface JavaEmulatorCommand {
   binary: string;
   args: string[];
   optionalArgs: string[];
+  joinArgs: boolean;
 }
 
-export interface JavaEmulatorDetails {
-  name: string;
-  instance: ChildProcess | null;
-  stdout: any | null;
+export interface JavaEmulatorOptions {
   cacheDir: string;
   remoteUrl: string;
   expectedSize: number;
   expectedChecksum: string;
-  localPath: string;
   namePrefix: string;
 }
 
+export abstract class JavaEmulatorDetails {
+  instance: ChildProcess | null = null;
+  stdout: any | null = null;
+
+  constructor(public name: string, public opts: JavaEmulatorOptions) {}
+
+  abstract get downloadPath(): string;
+  abstract get binaryPath(): string;
+  abstract get unzipDir(): string | undefined;
+}
+
+export class JarEmulatorDetails extends JavaEmulatorDetails {
+  constructor(name: string, private localPath: string, opts: JavaEmulatorOptions) {
+    super(name, opts);
+  }
+
+  get downloadPath(): string {
+    return this.localPath;
+  }
+
+  get binaryPath(): string {
+    return this.localPath;
+  }
+
+  get unzipDir(): string | undefined {
+    return undefined;
+  }
+}
+
+export class ZipEmulatorDetails extends JavaEmulatorDetails {
+  constructor(
+    name: string,
+    private localPath: string,
+    private _unzipDir: string,
```

Comment:
I cringe when I see underscore-prefixed variables. Can this just be `internalUnzipDir` or something? see the TS style guide for more... Also, it has discussions about getters that's interesting (go/ts-style)